File: examples/python/ex09_agent_workflow/main.py
"""09 Agent Workflow — tool-calling agent with a back-edge inside @graph.

1.0.0 migration note: pre-1.0 versions used @graph(until="done == True",
max_iterations=10) as the loop control. That surface is gone; the loop is
now expressed via a back-edge and the Phase 3 cycle-rewrite pass builds a
hidden _GraphLoop under the hood. The max-iterations cap defaults to 1000
on synthesized loops — the branch is your primary exit.


Requires ``OPENAI_API_KEY`` in ``.env`` and ``llm:gpt-4o-mini`` in
``resources.yaml``. Run from this directory:

    uv sync
    cp .env.example .env  # fill in OPENAI_API_KEY
    uv run python main.py
"""

from __future__ import annotations

import asyncio
import json

import operonx
from operonx.core import END, PARENT, START, Operon, graph, op
from operonx.core.ops.flow.branch_op import if_
from operonx.providers import LLMOp

# ── Tool functions ──────────────────────────────────────────────────────


def calculator(expression: str) -> dict:
    try:
        return {"result": str(eval(expression, {"__builtins__": {}}, {}))}
    except Exception as e:
        return {"error": str(e)}


def search(query: str) -> dict:
    knowledge = {
        "python": "Python is a high-level programming language created by Guido van Rossum in 1991.",
        "operonx": "Operonx is an async workflow orchestration engine for GenAI applications.",
        "vietnam": "Vietnam is a country in Southeast Asia. Capital: Hanoi. Population: ~100 million.",
        "machine learning": "Machine learning is a subset of AI that learns patterns from data.",
    }
    q = query.lower()
    for key, value in knowledge.items():
        if key in q:
            return {"result": value}
    return {"result": "No information found."}


TOOLS = {"calculator": calculator, "search": search}

TOOL_DESCRIPTIONS = [
    {
        "type": "function",
        "function": {
            "name": "calculator",
            "description": "Evaluate mathematical expressions. Example: '25 * 4 + 100'",
            "parameters": {
                "type": "object",
                "properties": {
                    "expression": {"type": "string", "description": "Math expression to evaluate"}
                },
                "required": ["expression"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "search",
            "description": "Search for factual information about a topic.",
            "parameters": {
                "type": "object",
                "properties": {"query": {"type": "string", "description": "Search query"}},
                "required": ["query"],
            },
        },
    },
]


# ── Agent ops ───────────────────────────────────────────────────────────


@op
def init_agent(query: str):
    return {
        "messages": [
            {
                "role": "system",
                "content": "You are a helpful assistant with access to tools. Use them when needed.",
            },
            {"role": "user", "content": query},
        ],
        "done": False,
        "answer": "",
    }


@op
def process_response(content, tool_calls, messages):
    """Run tool calls when present, otherwise mark `done`."""
    new_messages = list(messages)
    assistant = {"role": "assistant", "content": content or ""}
    if tool_calls:
        assistant["tool_calls"] = tool_calls
    new_messages.append(assistant)

    if tool_calls:
        for tc in tool_calls:
            fn_name = tc["function"]["name"]
            args = json.loads(tc["function"]["arguments"])
            if fn_name in TOOLS:
                tool_result = TOOLS[fn_name](**args)
            else:
                tool_result = {"error": f"Unknown tool: {fn_name}"}
            new_messages.append(
                {
                    "role": "tool",
                    "tool_call_id": tc["id"],
                    "content": json.dumps(tool_result),
                }
            )
        return {"messages": new_messages, "done": False, "answer": ""}

    return {"messages": new_messages, "done": True, "answer": content or ""}


# ── Loop body + outer graph ─────────────────────────────────────────────


@graph
def agent_loop():
    """Repeat LLM → process until ``done`` is True. The back-edge from
    ``proc`` to ``llm`` is what makes this a loop; the Phase 3 rewrite
    pass turns it into a hidden ``_GraphLoop`` at build time."""
    PARENT.declare(messages=[], done=False, answer="")
    llm = LLMOp.of(
        resource="gpt-4o-mini",
        messages=PARENT["messages"],
        tools=TOOL_DESCRIPTIONS,
    )
    proc = process_response(
        content=llm["content"],
        tool_calls=llm["tool_calls"],
        messages=PARENT["messages"],
    )
    proc["messages"] >> PARENT["messages"]
    proc["done"] >> PARENT["done"]
    proc["answer"] >> PARENT["answer"]
    START >> llm >> proc >> if_(proc["done"] == True, END).else_(llm)  # noqa: E712


@graph
def agent(query):
    """init → loop until the LLM stops calling tools."""
    init = init_agent(query=query)
    loop = agent_loop()
    # Wire init's outputs into the loop's shared cells at run start.
    init["messages"] >> loop["messages"]
    init["done"] >> loop["done"]
    init["answer"] >> loop["answer"]
    loop["answer"] >> PARENT["answer"]
    START >> init >> loop >> END


async def main() -> None:
    operonx.bootstrap()

    queries = [
        ("calc", "What is 25 * 4 + 100?"),
        ("search", "Tell me about Python programming language."),
        ("combined", "What is 15 * 7, and also tell me about machine learning?"),
    ]
    for label, query in queries:
        try:
            g = agent(query=PARENT["query"])
            result = await Operon(g).run(inputs={"query": query})
            print(f"[{label}] {result.get('answer', result)}")
        except Exception as e:
            print(f"[{label}] error: {e!r}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/rust/ex09_agent_workflow/src/main.rs
//! 09 Agent Workflow — Rust-side demo.
//!
//! Tool-calling agent mirroring `examples/python/ex09_agent_workflow/main.py`.
//! Only the plain `#[op]`s are declared; `LLMOp` is runtime-built.
//!
//! ⚠️  **Rust-runtime limitation** (today, v0.6.2): `@graph.loop` produces
//! a nested `OpType::Graph` with loop config that the Rust scheduler
//! currently returns empty for, so the agent will not iterate. The Python
//! side is fully functional.

use operonx::{op, BootstrapOpts, Operon};
use serde_json::Value;

#[op(name = "init_agent")]
fn init_agent(query: String) -> Value {
    serde_json::json!({
        "messages": [
            {
                "role": "system",
                "content": "You are a helpful assistant with access to tools. Use them when needed.",
            },
            {"role": "user", "content": query},
        ],
        "done": false,
        "answer": "",
    })
}

fn calculator(expr: &str) -> Value {
    // Stub — production needs a safe evaluator.
    serde_json::json!({ "result": format!("<computed:{expr}>") })
}

fn search(query: &str) -> Value {
    let q = query.to_lowercase();
    let knowledge = [
        ("python", "Python is a high-level programming language created by Guido van Rossum in 1991."),
        ("operonx", "Operonx is an async workflow orchestration engine for GenAI applications."),
        ("vietnam", "Vietnam is a country in Southeast Asia. Capital: Hanoi. Population: ~100 million."),
        ("machine learning", "Machine learning is a subset of AI that learns patterns from data."),
    ];
    for (k, v) in knowledge.iter() {
        if q.contains(k) {
            return serde_json::json!({ "result": v });
        }
    }
    serde_json::json!({ "result": "No information found." })
}

#[op(name = "process_response")]
fn process_response(content: Value, tool_calls: Value, messages: Value) -> Value {
    let mut new_messages: Vec<Value> = messages.as_array().cloned().unwrap_or_default();
    let calls: Vec<Value> = tool_calls.as_array().cloned().unwrap_or_default();

    let content_str = content.as_str().unwrap_or("").to_string();
    let mut assistant = serde_json::json!({
        "role": "assistant",
        "content": content_str.clone(),
    });
    if !calls.is_empty() {
        assistant["tool_calls"] = Value::Array(calls.clone());
    }
    new_messages.push(assistant);

    if !calls.is_empty() {
        for call in &calls {
            let func_name = call
                .pointer("/function/name")
                .and_then(Value::as_str)
                .unwrap_or("")
                .to_string();
            let args_raw = call
                .pointer("/function/arguments")
                .and_then(Value::as_str)
                .unwrap_or("{}");
            let args: Value = serde_json::from_str(args_raw).unwrap_or(serde_json::json!({}));

            let result = match func_name.as_str() {
                "calculator" => {
                    let expr = args.get("expression").and_then(Value::as_str).unwrap_or("");
                    calculator(expr)
                }
                "search" => {
                    let q = args.get("query").and_then(Value::as_str).unwrap_or("");
                    search(q)
                }
                other => serde_json::json!({ "error": format!("Unknown tool: {other}") }),
            };

            new_messages.push(serde_json::json!({
                "role": "tool",
                "tool_call_id": call.get("id").cloned().unwrap_or(Value::Null),
                "content": serde_json::to_string(&result).unwrap_or_else(|_| "{}".to_string()),
            }));
        }
        serde_json::json!({ "messages": new_messages, "done": false, "answer": "" })
    } else {
        serde_json::json!({ "messages": new_messages, "done": true, "answer": content_str })
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let here = std::path::Path::new(env!("CARGO_MANIFEST_DIR"));

    let opts = BootstrapOpts::new().resources(here.join("resources.yaml"));
    operonx::bootstrap(opts);

    let graph_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("graph.json"))?)?;
    let inputs_bundle: Value =
        serde_json::from_slice(&std::fs::read(here.join("inputs.json"))?)?;

    for name in ["calc", "search", "combined"] {
        let graph_v = graph_bundle
            .get(name)
            .ok_or_else(|| format!("graph.json missing `{name}` entry"))?;
        let graph_json = serde_json::to_string(graph_v)?;

        let inputs_obj = inputs_bundle
            .get(name)
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();

        let engine = Operon::builder(&graph_json).auto_register().build()?;
        match engine.run_json(inputs_obj, None, None, None) {
            Ok(r) => println!("[{name}] {r}"),
            Err(e) => println!("[{name}] error: {e}"),
        }
    }

    Ok(())
}
